File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.m
//
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 



#import "AudioTrackPlayer.h"
#import "AudioTrackPlayer+Private.h"

#import "UIImage+ImageUtilities.h"

@import WireCommonComponents;
@import WireUtilities;
@import WireSyncEngine;
#import "Wire-Swift.h"

@import AVFoundation;
@import MediaPlayer;

@implementation AudioTrackPlayer

- (void)dealloc
{
    self.audioTrack = nil;
    
    [self.avPlayer removeObserver:self forKeyPath:@"status"];
    [self.avPlayer removeObserver:self forKeyPath:@"rate"];
    [self.avPlayer removeObserver:self forKeyPath:@"currentItem"];
    
    [self setIsRemoteCommandCenterEnabled:NO];
}

- (void)loadTrack:(NSObject<AudioTrack> *)track sourceMessage:(id<ZMConversationMessage>)sourceMessage completionHandler:(void(^)(BOOL loaded, NSError *error))completionHandler
{
    _progress = 0;
    self.audioTrack = track;
    self.sourceMessage = sourceMessage;
    self.loadAudioTrackCompletionHandler = completionHandler;
    
    if (self.avPlayer == nil) {
        self.avPlayer = [AVPlayer playerWithURL:track.streamURL];
        [self.avPlayer addObserver:self forKeyPath:@"status" options:NSKeyValueObservingOptionNew context:NULL];
        [self.avPlayer addObserver:self forKeyPath:@"rate" options:NSKeyValueObservingOptionNew context:NULL];
        [self.avPlayer addObserver:self forKeyPath:@"currentItem" options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionOld context:NULL];
    } else {

        [self.avPlayer replaceCurrentItemWithPlayerItem:[AVPlayerItem playerItemWithURL:track.streamURL]];

        if (self.avPlayer.status == AVPlayerStatusReadyToPlay) {
            self.loadAudioTrackCompletionHandler(YES, nil);
        }
    }
    
    [[NSNotificationCenter defaultCenter] removeObserver:self name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(itemDidPlayToEndTime:) name:AVPlayerItemDidPlayToEndTimeNotification object:self.avPlayer.currentItem];
    
    if (self.timeObserverToken != nil) {
        [self.avPlayer removeTimeObserver:self.timeObserverToken];
    }
    
    __weak typeof(self) weakSelf = self;
    self.timeObserverToken = [self.avPlayer addPeriodicTimeObserverForInterval:CMTimeMake(1, 60) queue:dispatch_get_main_queue() usingBlock:^(CMTime time) {
        CMTimeRange itemRange = CMTimeRangeMake(CMTimeMake(0, 1), weakSelf.avPlayer.currentItem.asset.duration);
        CMTimeRange normalizedRange = CMTimeRangeMake(CMTimeMake(0, 1), CMTimeMake(1, 1));
        CMTime normalizedTime = CMTimeMapTimeFromRangeToRange(time, itemRange, normalizedRange);
        weakSelf.progress = CMTimeGetSeconds(normalizedTime);
    }];
    
    ZMUserSession *userSession = [ZMUserSession sharedSession];
    if (userSession != nil) {
        self.messageObserverToken = [MessageChangeInfo addObserver:self forMessage:sourceMessage userSession:[ZMUserSession sharedSession]];
    }
}

- (void)setIsRemoteCommandCenterEnabled:(BOOL)enabled
{
    MPRemoteCommandCenter *commandCenter = [MPRemoteCommandCenter sharedCommandCenter];
    
    if (!enabled) {
        [commandCenter.playCommand removeTarget:self.playHandler];
        [commandCenter.pauseCommand removeTarget:self.pauseHandler];
        [commandCenter.nextTrackCommand removeTarget:self.nextTrackHandler];
        [commandCenter.previousTrackCommand removeTarget:self.previousTrackHandler];
        return;
    }
    
    ZM_WEAK(self);
    self.pauseHandler = [commandCenter.pauseCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent *event) {
        ZM_STRONG(self);
        if (self.avPlayer.rate > 0) {
            [self pause];
            return MPRemoteCommandHandlerStatusSuccess;
        } else {
            return MPRemoteCommandHandlerStatusCommandFailed;
        }
    }];

    self.playHandler = [commandCenter.playCommand addTargetWithHandler:^MPRemoteCommandHandlerStatus(MPRemoteCommandEvent *event) {
        ZM_STRONG(self);
        if (self.audioTrack == nil) {
            return MPRemoteCommandHandlerStatusNoSuchContent;
        }

        if (self.avPlayer.rate == 0) {
            [self play];
            return MPRemoteCommandHandlerStatusSuccess;
        } else {
            return MPRemoteCommandHandlerStatusCommandFailed;
        }
    }];
}

- (NSTimeInterval)elapsedTime {
    CMTime time = self.avPlayer.currentTime;
    if (CMTIME_IS_VALID(time)) {
        return time.value / time.timescale;
    }
    return 0;
}

- (CGFloat)duration
{
    return CMTimeGetSeconds(self.avPlayer.currentItem.asset.duration);
}

- (BOOL)isPlaying
{
    return self.avPlayer.rate > 0 && self.avPlayer.error == nil;
}

- (void)play
{
    if (self.state == MediaPlayerStateCompleted) {
        [self.avPlayer seekToTime:CMTimeMake(0, 1)];
    }
    
    [self.avPlayer play];
}

- (void)pause
{
    [self.avPlayer pause];
}

- (void)stop
{
    [self.avPlayer pause];
    [self.avPlayer replaceCurrentItemWithPlayerItem:nil];
    self.audioTrack = nil;
    self.messageObserverToken = nil;
    _sourceMessage = nil;
}

- (NSString *)title
{
    return self.audioTrack.title;
}

+ (NSSet *)keyPathsForValuesAffectingPlaying
{
    return [NSSet setWithObjects:@"avPlayer.rate", nil];
}

+ (NSSet *)keyPathsForValuesAffectingError
{
    return [NSSet setWithObjects:@"avPlayer.error", nil];
}

- (void)setAudioTrack:(NSObject<AudioTrack> *)audioTrack
{
    if (_audioTrack == audioTrack) {
        return;
    }
    
    if (_audioTrack != nil) {
        [_audioTrack removeObserver:self forKeyPath:@"status"];
    }
    _audioTrack = audioTrack;
    if (_audioTrack != nil) {
        [_audioTrack addObserver:self forKeyPath:@"status" options:NSKeyValueObservingOptionNew context:NULL];
    }
}

#pragma mark - ZMMessageObserver

- (void)messageDidChange:(MessageChangeInfo *)changeInfo
{
    if (changeInfo.message.hasBeenDeleted) {
        [self stop];
    }
}

#pragma mark - KVO observer

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (object == self.avPlayer.currentItem && [keyPath isEqualToString:@"status"]) {
        if (self.avPlayer.currentItem.status == AVPlayerItemStatusFailed) {
            self.state = MediaPlayerStateError;
            self.audioTrack.failedToLoad = YES;
            [self.mediaPlayerDelegate mediaPlayer:self didChangeToState:self.state];
        }
    }
    
    if (object == self.avPlayer && [keyPath isEqualToString:@"status"]) {
        if (self.avPlayer.status == AVPlayerStatusReadyToPlay) {
            self.loadAudioTrackCompletionHandler(YES, nil);
        }
        else if (self.avPlayer.status == AVPlayerStatusFailed) {
            self.loadAudioTrackCompletionHandler(NO, self.avPlayer.error);
        }
    }
    
    if (object == self.avPlayer && [keyPath isEqualToString:@"rate"]) {
        
        if (self.avPlayer.rate > 0) {
            self.state = MediaPlayerStatePlaying;
            [self.mediaPlayerDelegate mediaPlayer:self didChangeToState:self.state];
        } else if (self.state != MediaPlayerStateCompleted) {
            self.state = MediaPlayerStatePaused;
            [self.mediaPlayerDelegate mediaPlayer:self didChangeToState:self.state];
        }
        
        [self updateNowPlayingState];
    }
    
    if (object == self.avPlayer && [keyPath isEqualToString:@"currentItem"]) {
     
        if (self.avPlayer.currentItem == nil) {
            [self setIsRemoteCommandCenterEnabled:NO];
            [self clearNowPlayingState];
            self.state = MediaPlayerStateCompleted;
            [self.mediaPlayerDelegate mediaPlayer:self didChangeToState:self.state];
        } else {
            [self setIsRemoteCommandCenterEnabled:YES];
            [self populateNowPlayingState];
        }
    }
    
}

#pragma mark - MPNowPlayingInfoCenter

- (void)clearNowPlayingState
{
    MPNowPlayingInfoCenter* info = [MPNowPlayingInfoCenter defaultCenter];
    info.nowPlayingInfo = nil;
    self.nowPlayingInfo = nil;
}

- (void)updateNowPlayingState
{
    NSMutableDictionary *newInfo = [self.nowPlayingInfo mutableCopy];
    [newInfo setObject:@(self.elapsedTime) forKey:MPNowPlayingInfoPropertyElapsedPlaybackTime];
    [newInfo setObject:@(self.avPlayer.rate) forKey:MPNowPlayingInfoPropertyPlaybackRate];
 
    MPNowPlayingInfoCenter* info = [MPNowPlayingInfoCenter defaultCenter];
    info.nowPlayingInfo = newInfo;
    self.nowPlayingInfo = newInfo;
}
        
#pragma mark AVPlayer notifications

- (void)itemDidPlayToEndTime:(NSNotification *)notification
{
    // AUDIO-557 workaround for AVSMediaManager trying to pause already paused tracks.
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self clearNowPlayingState];
        self.state = MediaPlayerStateCompleted;
        [self.mediaPlayerDelegate mediaPlayer:self didChangeToState:self.state];
    });
}

@end

File: wire-ios/Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import MediaPlayer
import WireSyncEngine

// MARK: - MediaPlayerState

/// These enums represent the state of the current media in the player.

enum MediaPlayerState: Int {
    case ready = 0
    case playing
    case paused
    case completed
    case error
}

// MARK: - MediaPlayer

protocol MediaPlayer: AnyObject {
    var title: String? { get }
    var sourceMessage: ZMConversationMessage? { get }
    var state: MediaPlayerState? { get }
    func play()
    func pause()
    func stop()
}

typealias AudioTrackCompletionHandler = (_ loaded: Bool, _ error: Error?) -> Void

// MARK: - AudioTrackPlayerDelegate

protocol AudioTrackPlayerDelegate: AnyObject {
    func stateDidChange(_ audioTrackPlayer: AudioTrackPlayer, state: MediaPlayerState?)
    func progressDidChange(_ audioTrackPlayer: AudioTrackPlayer, progress: Double)
}

// MARK: - AudioTrackPlayer

final class AudioTrackPlayer: NSObject, MediaPlayer {
    private let userSession: UserSession

    private var avPlayer: AVPlayer?
    private var timeObserverToken: Any?
    private var messageObserverToken: NSObjectProtocol?
    private var loadAudioTrackCompletionHandler: AudioTrackCompletionHandler?

    private(set) var playing = false

    weak var audioTrackPlayerDelegate: AudioTrackPlayerDelegate?
    weak var mediaPlayerDelegate: MediaPlayerDelegate?

    var state: MediaPlayerState? {
        didSet {
            audioTrackPlayerDelegate?.stateDidChange(self, state: state)
            if let state {
                mediaPlayerDelegate?.mediaPlayer(self, didChangeTo: state)
            }
        }
    }

    var sourceMessage: ZMConversationMessage?
    private var nowPlayingInfo: [String: Any]?
    private var playHandler: Any?
    private var pauseHandler: Any?
    private var nextTrackHandler: Any?
    private var previousTrackHandler: Any?

    private var playerStatusObserver: NSKeyValueObservation?
    private var playerRateObserver: NSKeyValueObservation?
    private var playerCurrentItemObserver: NSKeyValueObservation?

    private(set) var audioTrack: AudioTrack?

    private(set) var progress: Double = 0 {
        didSet {
            audioTrackPlayerDelegate?.progressDidChange(self, progress: progress)
        }
    }

    var duration: Double {
        if let duration = avPlayer?.currentItem?.asset.duration {
            return CMTimeGetSeconds(duration)
        }
        return 0
    }

    var elapsedTime: TimeInterval {
        guard let time = avPlayer?.currentTime() else { return 0 }

        if CMTIME_IS_VALID(time) {
            return TimeInterval(time.value) / TimeInterval(time.timescale)
        }

        return 0
    }

    /// Start the currently loaded/paused track.
    func play() {
        if state == .completed {
            avPlayer?.seek(to: CMTimeMake(value: 0, timescale: 1))
        }

        avPlayer?.play()
    }

    /// Pause the currently playing track.
    func pause() {
        avPlayer?.pause()
    }

    init(userSession: UserSession) {
        self.userSession = userSession
        super.init()
    }

    deinit {
        setIsRemoteCommandCenterEnabled(false)
    }

    func load(
        _ track: AudioTrack,
        sourceMessage: ZMConversationMessage,
        completionHandler: AudioTrackCompletionHandler? = nil
    ) {
        progress = 0
        audioTrack = track
        self.sourceMessage = sourceMessage
        loadAudioTrackCompletionHandler = completionHandler

        if let streamURL = track.streamURL {
            if let avPlayer {
                avPlayer.replaceCurrentItem(with: AVPlayerItem(url: streamURL))

                if avPlayer.status == .readyToPlay {
                    loadAudioTrackCompletionHandler?(true, nil)
                }
            } else {
                avPlayer = AVPlayer(url: streamURL)

                playerStatusObserver = avPlayer?.observe(\AVPlayer.status, options: [.new]) { [weak self] _, _ in
                    self?.playStatusChanged()
                }

                playerRateObserver = avPlayer?.observe(\AVPlayer.rate, options: [.new]) { [weak self] _, _ in
                    self?.playRateChanged()
                }

                playerCurrentItemObserver = avPlayer?.observe(\AVPlayer.currentItem, options: [
                    .new,
                    .initial,
                    .old,
                ]) { [weak self] _, _ in
                    self?.playCurrentItemChanged()
                }
            }
        } else {
            // For testing only! streamURL is nil in tests.
            avPlayer = AVPlayer()
        }

        NotificationCenter.default.removeObserver(self, name: .AVPlayerItemDidPlayToEndTime, object: nil)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(itemDidPlay(toEndTime:)),
            name: .AVPlayerItemDidPlayToEndTime,
            object: avPlayer?.currentItem
        )

        if let timeObserverToken {
            avPlayer?.removeTimeObserver(timeObserverToken)
        }

        timeObserverToken = avPlayer?.addPeriodicTimeObserver(
            forInterval: CMTimeMake(value: 1, timescale: 60),
            queue: DispatchQueue.main,
            using: { [weak self] time in
                guard let self, let duration = avPlayer?.currentItem?.asset.duration else { return }

                let itemRange = CMTimeRangeMake(start: CMTimeMake(value: 0, timescale: 1), duration: duration)

                let normalizedRange = CMTimeRangeMake(
                    start: CMTimeMake(value: 0, timescale: 1),
                    duration: CMTimeMake(value: 1, timescale: 1)
                )

                let normalizedTime = CMTimeMapTimeFromRangeToRange(time, fromRange: itemRange, toRange: normalizedRange)

                progress = CMTimeGetSeconds(normalizedTime)
            }
        )

        messageObserverToken = userSession.addMessageObserver(
            self,
            for: sourceMessage
        )
    }

    private func playRateChanged() {
        if let avPlayer, avPlayer.rate > 0 {
            state = .playing
        } else {
            state = .paused
        }

        updateNowPlayingState()
    }

    private func playCurrentItemChanged() {
        if avPlayer?.currentItem == nil {
            setIsRemoteCommandCenterEnabled(false)
            clearNowPlayingState()
            state = .completed
        } else {
            setIsRemoteCommandCenterEnabled(true)
            populateNowPlayingState()
        }
    }

    private func playStatusChanged() {
        if avPlayer?.currentItem?.status == .failed {
            audioTrack?.failedToLoad = true
            state = .error
        }

        guard let status = avPlayer?.status else { return }

        switch status {
        case .readyToPlay:
            loadAudioTrackCompletionHandler?(true, nil)
        case .failed:
            loadAudioTrackCompletionHandler?(false, avPlayer?.error)
        default:
            break
        }
    }

    func setIsRemoteCommandCenterEnabled(_ enabled: Bool) {
        let commandCenter = MPRemoteCommandCenter.shared()

        if !enabled {
            commandCenter.playCommand.removeTarget(playHandler)
            commandCenter.pauseCommand.removeTarget(pauseHandler)
            commandCenter.nextTrackCommand.removeTarget(nextTrackHandler)
            commandCenter.previousTrackCommand.removeTarget(previousTrackHandler)
            return
        }

        pauseHandler = commandCenter.pauseCommand.addTarget(handler: { [weak self] _ in
            if let avPlayer = self?.avPlayer, avPlayer.rate > 0 {
                self?.pause()
                return .success
            } else {
                return .commandFailed
            }
        })

        playHandler = commandCenter.playCommand.addTarget(handler: { [weak self] _ in
            if self?.audioTrack == nil {
                return .noSuchContent
            }

            if self?.avPlayer?.rate == 0 {
                self?.play()
                return .success
            } else {
                return .commandFailed
            }
        })
    }

    var isPlaying: Bool {
        if let avPlayer, avPlayer.rate > 0, avPlayer.error == nil {
            return true
        }

        return false
    }

    func stop() {
        avPlayer?.pause()
        avPlayer?.replaceCurrentItem(with: nil)
        audioTrack = nil
        messageObserverToken = nil
        sourceMessage = nil
    }

    var title: String? {
        audioTrack?.title
    }

    // MARK: - MPNowPlayingInfoCenter

    private func clearNowPlayingState() {
        let info = MPNowPlayingInfoCenter.default()
        info.nowPlayingInfo = nil
        nowPlayingInfo = nil
    }

    private func updateNowPlayingState() {
        var newInfo = nowPlayingInfo
        newInfo?[MPNowPlayingInfoPropertyElapsedPlaybackTime] = NSNumber(value: elapsedTime)
        if let rate = avPlayer?.rate {
            newInfo?[MPNowPlayingInfoPropertyPlaybackRate] = NSNumber(value: rate)
        }

        let info = MPNowPlayingInfoCenter.default()
        info.nowPlayingInfo = newInfo
        nowPlayingInfo = newInfo
    }

    // MARK: AVPlayer notifications

    @objc
    private func itemDidPlay(toEndTime notification: Notification?) {
        // AUDIO-557 workaround for AVSMediaManager trying to pause already paused tracks.
        delay(0.1) { [weak self] in
            guard let self else { return }

            clearNowPlayingState()
            state = .completed
        }
    }

    // MARK: - MPNowPlayingInfoCenter

    func populateNowPlayingState() {
        let playbackDuration: NSNumber = if let duration: CMTime = avPlayer?.currentItem?.asset.duration {
            NSNumber(value: CMTimeGetSeconds(duration))
        } else {
            0
        }

        let nowPlayingInfo: [String: Any] = [
            MPMediaItemPropertyTitle: audioTrack?.title ?? "",
            MPMediaItemPropertyArtist: audioTrack?.author ?? "",
            MPNowPlayingInfoPropertyPlaybackRate: NSNumber(value: avPlayer?.rate ?? 0),
            MPMediaItemPropertyPlaybackDuration: playbackDuration,
        ]

        MPNowPlayingInfoCenter.default().nowPlayingInfo = nowPlayingInfo
        self.nowPlayingInfo = nowPlayingInfo
    }
}

// MARK: ZMMessageObserver

extension AudioTrackPlayer: ZMMessageObserver {
    func messageDidChange(_ changeInfo: MessageChangeInfo) {
        if changeInfo.message.hasBeenDeleted {
            stop()
        }
    }
}
